File: 007_Programming_Assignment_Dynamic_Programming/placing_parentheses/placing_parentheses.cpp
#include <iostream>
#include <cassert>
#include <string>
#include <vector>
#include <array>

using std::vector;
using std::string;
using std::max;
using std::min;

struct MinAndMax_t {
  int min;
  int max;
};

long long eval(long long a, long long b, char op) {
  if (op == '*') {
    return a * b;
  } else if (op == '+') {
    return a + b;
  } else if (op == '-') {
    return a - b;
  } else {
    assert(0);
  }
}

long long get_maximum_value(const string &exp) {
  //write your code here
  long long count_of_numbers = exp.length();
  if (exp.length() % 2 != 0) {
    count_of_numbers += 1;
  }

  // initialize min and max matrices
  //long long min_vals[][] = new long long[count_of_numbers][count_of_numbers];
  std::array<std::array<int, 3>, 3> = new array(3, 3);

  long long i;
  for (i = 0; i < count_of_numbers; ++i) {

  }

  return 0;
}

int main() {
  string s;
  std::cin >> s;
  std::cout << get_maximum_value(s) << '\n';
}

File: 007_Programming_Assignment_Dynamic_Programming/placing_parentheses/placing_parentheses.py
# Uses python3
import math

def evalt(a, op, b):
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    elif op == '*':
        return a * b
    else:
        assert False

def get_maximum_value(dataset):
    #write your code here
    n = int((len(dataset) / 2)) + 1
    # generate new string d containing numbers without operators
    nums = list(map(int,''.join(c for c in dataset if c not in '+-*')))
    # generate new list with of all operators
    ops = list(''.join(c for c in dataset if c not in '0123456789'))

    min_matrix = [[0 for x in range(n)] for y in range(n)]
    max_matrix = [[0 for x in range(n)] for y in range(n)]

    for i in range(n):
        min_matrix[i][i] = nums[i]
        max_matrix[i][i] = nums[i]

    for s in range(1,n): # start range at 1 to shift to 1st calculation col
        for i in range(n-s):
            j = i + s
            ########### MinAndMax
            max_num = -math.inf
            min_num = math.inf
            for k in range(i, j):
                fa = evalt(max_matrix[i][k], ops[k], max_matrix[k+1][j])
                fb = evalt(max_matrix[i][k], ops[k], min_matrix[k+1][j])
                fc = evalt(min_matrix[i][k], ops[k], max_matrix[k+1][j])
                fd = evalt(min_matrix[i][k], ops[k], min_matrix[k+1][j])
                min_num = min(min_num, fa, fb, fc, fd)
                max_num = max(max_num, fa, fb, fc, fd)
            ###########
            min_matrix[i][j] = min_num
            max_matrix[i][j] = max_num
    return max_matrix[0][n-1]

if __name__ == "__main__":
    print(get_maximum_value(input()))
